empiric-package/empiric/core/mixins/nonce.py

```diff
@@ -0,0 +1,98 @@
+from typing import Dict, Literal, Optional, Union
+
+from starknet_py.net.client import Client
+from starknet_py.net.client_models import TransactionReceipt, TransactionStatus
+
+
+class NonceMixin:
+    client: Client
+    account_contract_address: Optional[int]
+    nonce_status: Dict[int, TransactionStatus] = dict()
+    nonce_dict: Dict[int, int] = dict()
+    pending_nonce: Optional[int] = None
+
+    async def _get_nonce(self) -> int:
+        latest_nonce = await self.get_nonce()
+        self.update_nonce_dict()
+        self.cleanup_nonce_dict()
+
+        if self.pending_nonce:
+            return self.pending_nonce
+        if self.nonce_status:
+            return max(self.nonce_status) + 1
+
+        return latest_nonce
+
+    def cleanup_nonce_dict(self):
+        nonce_seq = [
+            x
+            for x in self.nonce_dict
+            if self.nonce_status[x]
+            in [TransactionStatus.ACCEPTED_ON_L1, TransactionStatus.ACCEPTED_ON_L2]
+        ]
+        if nonce_seq:
+            max_accepted = max(nonce_seq)
+            for nonce in list(self.nonce_dict):
+                if nonce <= max_accepted:
+                    del self.nonce_dict[nonce]
+                    if nonce in self.nonce_status:
+                        del self.nonce_status[nonce]
+            if self.pending_nonce and self.pending_nonce < max_accepted:
+                self.pending_nonce = None
+
+    async def track_nonce(
+        self,
+        nonce,
+        transaction_hash,
+    ):
+        self.nonce_dict[nonce] = transaction_hash
+        self.nonce_status[nonce] = await self.get_status(transaction_hash)
+
+        nonce_min = min(self.nonce_dict)
+        nonce_max = max(self.nonce_dict)
+        for i in range(nonce_min, nonce_max + 1):
+            if i not in self.nonce_dict:
+                self.pending_nonce = i
+                break
+
+    async def update_nonce_dict(
+        self,
+    ):
+        for nonce in list(self.nonce_dict):
+            self.nonce_status[nonce] = await self.get_status(self.nonce_dict[nonce])
```

Comment:
does it make sense to try to parallelize this?

Reply:
there shouldn't be that many nonces pending at any given time so I wouldn't expect it to matter too much.  we could if we were concerned with performance.  Typically the response time is less than a second.

Reply:
Sounds good; let's leave for now but in the future this is low-hanging fruit for reducing latency